contracts/RenExBalances.sol

```diff
@@ -7,157 +7,107 @@ import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
-    owner, intended to be the RenEx settlement contract
-    @param _trader the address of the trader
-    @param _token the token's address
-    @param _value the number of tokens to increment the balance by (in the token's smallest unit)
-    */
-    function incrementBalance(address _trader, address _token, uint256 _value) public onlyRenExSettlementContract {
-        privateIncrementBalance(_trader, ERC20(_token), _value);
-    }
-
-    function privateIncrementBalance(address _trader, ERC20 _token, uint256 _value) private {
-        // Check if it's the first time the trader
-        if (!activeTraderToken[_trader][_token]) {
-            activeTraderToken[_trader][_token] = true;
-            traderTokens[_trader].push(_token);
-        }
-
-        traderBalances[_trader][_token] = traderBalances[_trader][_token].add(_value);
-
-        emit BalanceIncreased(_trader, _token, _value);
-    }
-
-    /**
-    @notice Decrements a trader's balance of a token - can only be called by the
-    owner, intended to be the RenEx settlement contract
-    @param _trader the address of the trader
-    @param _token the token's address
-    @param _value the number of tokens to decrement the balance by (in the token's smallest unit)
-    */
-    function decrementBalanceWithFee(address _trader, address _token, uint256 _value, uint256 _fee, address feePayee)
-    public onlyRenExSettlementContract {
-        require(traderBalances[_trader][_token] >= _fee, "insufficient funds for fee");
+    /// @notice Transfer a token value from one trader to another, transferring
+    /// a fee to the RewardVault. Can only be called by the RenExSettlement
+    /// contract.
+    ///
+    /// @param _traderFrom The address of the trader to decrement the balance of.
+    /// @param _traderTo The address of the trader to increment the balance of.
+    /// @param _token The token's address.
+    /// @param _value The number of tokens to decrement the balance by (in the
+    ///        token's smallest unit).
+    /// @param _fee The fee amount to forward on to the RewardVault.
+    /// @param _feePayee The recipient of the fee.
+    function transferBalanceWithFee(address _traderFrom, address _traderTo, address _token, uint256 _value, uint256 _fee, address _feePayee)
+    external onlyRenExSettlementContract {
+        require(traderBalances[_traderFrom][_token] >= _fee, "insufficient funds for fee");
 
         if (address(_token) == ETHEREUM) {
-            rewardVaultContract.deposit.value(_fee)(feePayee, ERC20(_token), _fee);
+            rewardVaultContract.deposit.value(_fee)(_feePayee, ERC20(_token), _fee);
         } else {
             ERC20(_token).approve(rewardVaultContract, _fee);
-            rewardVaultContract.deposit(feePayee, ERC20(_token), _fee);
+            rewardVaultContract.deposit(_feePayee, ERC20(_token), _fee);
+        }
+        privateDecrementBalance(_traderFrom, ERC20(_token), _value + _fee);
+        if (_value > 0) {
```

Comment:
We can get rid of this `if`. It's a `uint256` so it can't be less than 0, and so only saves gas if the value is 0. Since this is usually not the case, best to remove it.

Reply:
Good point. It's checked because it will be 0 for atomic swap fees. We can change it to (_value != 0) which is more gas-efficient, or remove the check.

Reply:
Let's change it to `!= 0`

Reply:
Actually`>0` compiles to just `gt` (3 gas). `!= 0` compiles to `eq, iszero` (6 gas). The whole check doesn't cost too much more. I'll leave it for now.